metabase/metabase
# Allow joins on multiple fields in the notebook editor

This PR updates the `Join` class and notebook editor to support joining tables on multiple fields. The underlying MBQL is translated into SQL like:

```sql
SELECT * FROM Orders AS o
INNER JOIN Products AS p
ON o.PRODUCT_ID = p.ID AND o.CREATED_AT = p.CREATED_AT
```

`Join` class from `metabase-lib` used to have a single `condition` property storing the join dimension references (as MBQL). Let's say we want to join the Orders table with the Products table by Product ID. Join's `condition` in that case will look like the following:

```js
const condition = [
   "=", // eq operator means the following two field are expected to be equal
   [ "field", 3, null ], // left side field (ORDERS.PRODUCT_ID, the number is field's internal ID)
   [ "field", 14, { "join-alias": "Products" } ], // right side field (PRODUCTS.ID, the number is field's internal ID)
]
```

Now `Join` would also support more complex conditions. Let's say we now also want to join tables on the `CREATED_AT` field. The condition will now look like that:

```js
const condition = [
   "and",
   // The expressions below are the same as in the example above.
   // The whole idea is that we combine them using "and" operator
   [ "=", [ "field", 3, null ], [ "field", 14, { "join-alias": "Products" } ],
   [ "=", [ "field", 7, null ], [ "field", 19, { "join-alias": "Products" } ],
]
```

### To Verify

In order to test multiple joins, we're going to join the Orders and Products tables on Product ID and creation date (orders created in the same month with a related product).

1. Ask a question > Custom question > Sample Dataset > Orders
2. Join Products table on `Orders.PRODUCT_ID = Products.ID`
3. Summarize by "Count of rows"
4. Select "inner" join strategy (inner join will provide the most notable results)
4. Click "Visualize"
5. You should see there are 18760 records
6. Open the notebook again and extend add another join dimensions pair: `Orders.CREATED_AT` on `Products.CREATED_AT`
7. Click "Visualize"
8. You should see there are 415 records now. The number changed because `INNER JOIN` returned only records where the order's creation date matches the product's creation date (by month, not by exact date time)
9. If you clean up the summarize step (remove "Count of rows", you should see that for every row `Product ID` and `Product > ID` columns & `Created At` and `Product > Created At` match

### Demo


https://user-images.githubusercontent.com/17258145/131162557-dc8768c0-2e22-4c0a-87f9-826f54feee19.mov
